[MM-54367] Apply scope to the menu class

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

Related conversation: https://community.mattermost.com/core/pl/cfk9y3o8jpy9tn4dhz4cd4ioqr

I found that the Todo plugin's css affects the entire `Menu` component.
I applied scope to the `Menu` component.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes https://github.com/mattermost/mattermost/issues/24455
Jira https://mattermost.atlassian.net/browse/MM-54367

kyeongsoosoo commented:
I found that the `Button` and `Chip` component have the same bugs. (https://github.com/mattermost/mattermost-plugin-todo/blob/a309a7ab63f0ea0ec7b91e7f46b22045e84ee08e/webapp/src/widget/buttons/button.scss#L1)
Would it be better to fix them together in this PR?

larkox commented on webapp/src/widget/menu.scss:
```diff
@@ -125,8 +125,8 @@
-.Menu,
-.SubMenuOption .SubMenu {
+.TodoMenu,
+.SubTodoMenuOption .SubTodoMenu {
```
I see changes in the code for `Menu`->`TodoMenu`. Is there any change for `SubMenuOption`->`SubTodoMenuOption` needed?

kyeongsoosoo replied:
I think it's not needed now. I added it for future use-case.

mickmister replied:
If there is no class name being used in the components `SubTodoMenuOption` or `SubTodoMenu`, then I think they should be removed from the scss file to avoid confusion. Likely, [yagni](https://martinfowler.com/bliki/Yagni.html)

kyeongsoosoo replied:
@mickmister I removed the unused classnames.

mickmister commented on webapp/src/widget/menu.scss:
```diff
@@ -1,10 +1,10 @@
 .status-dropdown-menu {
-    > .Menu {
+    > .TodoMenu {
```
Unrelated to this PR, but it seems wrong to assume that the plugin's `TodoMenu` component will always be a next-level descendant of the element that has `status-dropdown-menu` class name

kyeongsoosoo replied:
I thought it would be okay because  `position` is applied separately to the TodoMenu classname.
https://github.com/mattermost/mattermost-plugin-todo/blob/a309a7ab63f0ea0ec7b91e7f46b22045e84ee08e/webapp/src/widget/menu.scss#L7-L10

mickmister replied:
Right, but `.status-dropdown-menu` is managed by the core webapp. At any point, the webapp could restructure itsrlf and do one of:
- Remove that class name altogether
- Change its component structure to have a component in between `.status-dropdown-menu` and the plugin's component

Both of these would make it so our plugin styles don't get used. The alternative is to not have the `>` in the selector

mickmister reviewed (approved):
Thanks for catching this and implementing the fix @kyeongsoosoo :+1:

The changes LGTM. I just have one non-blocking suggestion to remove the class names from the scss file that are not being used in the React code `SubTodoMenuOption` and `SubTodoMenu`

AayushChaudhary0001 reviewed (commented):
The issue has been solved in this PR. The UI does not get hampered through this fix, but the user is still able to resize the LHS while the channel menu is opened.
**Steps to reproduce:-**
1. Open the channel menu.
2. Try to resize the LHS.

The user is able to resize the LHS which contrasts the expected result in the issue.

![Screenshot from 2023-09-20 16-43-15](https://github.com/mattermost/mattermost-plugin-todo/assets/116162036/d0af25a8-2863-4bc9-9b77-e26c0222f44d)

mickmister commented:
@AayushChaudhary0001 I think the issue you point out may be unrelated to this PR. If the issue the PR intends to fix is fixed by this PR, can you please approve it? We can make a separate issue for the main repository if there's an unrelated issue

mickmister commented:
@kyeongsoosoo Heads up that there are some linting issues to resolve on the PR

kyeongsoosoo commented:
/update-branch

kyeongsoosoo commented on webapp/src/widget/menuWrapper.tsx:
```diff
@@ -19,7 +19,7 @@ const MenuWrapper = React.memo((props: Props) => {
 
     if (!Array.isArray(props.children) || props.children.length !== 2) {
-        throw new Error('MenuWrapper needs exactly 2 children');
+        throw new Error('TodoMenuWrapper needs exactly 2 children');
```
I reverted it back to `MenuWrapper` to keep it the same as the other components.

kyeongsoosoo commented:
@mickmister Could you help me re-run the test? I fixed the linting issue. 
The issue was on the server side. It's not PR related, but it was a minor change, so I fixed it.

mickmister commented:
@kyeongsoosoo Thanks for fixing that :+1:

AayushChaudhary0001 commented:
@mickmister  Yes, I am approving this PR and we can create a different issue for this, but in the description of the PR, the expected result in the mentioned ticket should be similar to what has been tested, so kindly update that.

mickmister commented:
@kyeongsoosoo Merging this, as it's obvious the issue that this plugin was causing is fixed. Is it your understanding that https://github.com/mattermost/mattermost/issues/24455 is solved, is there additional work to be done in the main repo to completely resolve the issue?

kyeongsoosoo commented:
@mickmister Sorry for the late reply. 
I think the ticket is solved. The issue mentioned above was not in the spec at the time of implementation, so I don't think it is a bug that should be fixed in this ticket.
